test/shopify_app/controller_concerns/token_exchange_test.rb

```diff
@@ -0,0 +1,203 @@
+# frozen_string_literal: true
+
+require "test_helper"
+require "action_controller"
+require "action_controller/base"
+
+class TokenExchangeController < ActionController::Base
+  include ShopifyApp::TokenExchange
+
+  around_action :activate_shopify_session
+
+  def index
+    render(plain: "OK")
+  end
+end
+
+class TokenExchangeControllerTest < ActionController::TestCase
+  tests TokenExchangeController
+
+  setup do
+    @shop = "my-shop.myshopify.com"
+    ShopifyApp::JWT.any_instance.stubs(:shopify_domain).returns(@shop)
+
+    @session_token = "this-is-a-session-token"
+    @session_token_in_header = "Bearer #{@session_token}"
+    request.headers["HTTP_AUTHORIZATION"] = @session_token_in_header
+
+    ShopifyApp::SessionRepository.shop_storage = ShopifyApp::InMemoryShopSessionStore
+    ShopifyApp::SessionRepository.user_storage = nil
+
+    @user = ShopifyAPI::Auth::AssociatedUser.new(
+      id: 1,
+      first_name: "Hello",
+      last_name: "World",
+      email: "Email",
+      email_verified: true,
+      account_owner: true,
+      locale: "en",
+      collaborator: false,
+    )
+    @online_session = ShopifyAPI::Auth::Session.new(
+      id: "online_session_1",
+      shop: @shop,
+      is_online: true,
+      associated_user: @user,
+    )
+    @offline_session = ShopifyAPI::Auth::Session.new(id: "offline_session_1", shop: @shop)
+
+    @offline_session_id = "offline_#{@shop}"
+    @online_session_id = "online_#{@user.id}"
+  end
+
+  test "Exchanges offline token when session doesn't exist" do
+    with_application_test_routes do
+      ShopifyAPI::Utils::SessionUtils.expects(:current_session_id).twice.with(
```

Comment:
I'd maybe go with a `stubs` here instead of an expectation. It doesn't seem essential to the test outcome that this gets called twice; that's an implementation detail. The thing we care about is that the `exchange_token` happens once, which you're expecting below. What do you think?